src/kcas.ml

```diff
@@ -370,59 +516,101 @@ module Xt = struct
   let update ~xt loc f = update loc (protect xt f) xt
   let swap ~xt l1 l2 = set ~xt l1 @@ exchange ~xt l2 @@ get ~xt l1
 
+  let to_blocking ~xt tx =
+    match tx ~xt with None -> Retry.later () | Some value -> value
+    [@@inline]
+
+  let to_nonblocking ~xt tx =
+    match tx ~xt with value -> Some value | exception Retry.Later -> None
+    [@@inline]
+
   let post_commit ~xt action =
     xt.post_commit <- Action.append action xt.post_commit
 
   let is_in_log ~xt loc =
     let x = loc.id in
     match xt.cass with
     | NIL -> false
-    | CASN (a, _, NIL, _) when x < a.id -> false
-    | CASN (a, _, _, NIL) when a.id < x -> false
-    | CASN (a, _, _, _) when Obj.magic a == loc -> true
+    | CASN { loc = a; lt = NIL; _ } when x < a.id -> false
+    | CASN { loc = a; gt = NIL; _ } when a.id < x -> false
+    | CASN { loc = a; _ } when Obj.magic a == loc -> true
     | cass -> (
         match splay ~hit_parent:true x cass with
-        | l, Hit (a, s), r ->
-            xt.cass <- CASN (a, s, l, r);
+        | lt, Hit (a, state), gt ->
+            xt.cass <- CASN { loc = a; state; lt; gt; awaiters = [] };
             Obj.magic a == loc
         | _, Miss, _ -> impossible ())
 
   type 'a tx = { tx : 'x. xt:'x t -> 'a } [@@unboxed]
 
   let call { tx } = tx [@@inline]
 
-  let attempt (mode : Mode.t) tx =
+  let rec add_awaiters awaiter casn = function
+    | NIL as cont -> cont
+    | CASN { loc; state; lt; gt; _ } as stop -> (
+        match if lt == NIL then lt else add_awaiters awaiter casn lt with
+        | NIL ->
+            if
+              add_awaiter loc
+                (if is_cmp casn state then eval state else state.before)
+                awaiter
+            then add_awaiters awaiter casn gt
+            else stop
+        | CASN _ as stop -> stop)
```

Comment:
`add_awaiters` signals that some location had already been updated by returning the `CASN _` descriptor of that location. This way the subsequent call of `remove_awaiters` (to prevent space leaks) can stop early at that same descriptor.